pinot-server/src/main/java/org/apache/pinot/server/starter/helix/HelixServerStarter.java

```diff
@@ -363,6 +363,7 @@ public void start()
 
     LOGGER.info("Initializing server instance and registering state model factory");
     Utils.logVersions();
+    ControllerLeaderLocator.create(_helixManager);
```

Comment:
ControllerLeaderLocator uses helixManager to get the controller resource status, so you need to have the helix manager connected before any leader locator functionality is invoked. So your use case will not work if the same race condition happens again.

Reply:
Here we only hold a reference to helixManager
since we set ServerInstance::start as a addPreConnectCallback, then the server will always start after helixManager connected.